src/main/java/bio/terra/service/snapshotbuilder/utils/CriteriaQueryBuilder.java

```diff
@@ -0,0 +1,211 @@
+import bio.terra.service.snapshotbuilder.query.FieldVariable;
+import bio.terra.service.snapshotbuilder.query.FilterVariable;
+import bio.terra.service.snapshotbuilder.query.Literal;
+import bio.terra.service.snapshotbuilder.query.Query;
+import bio.terra.service.snapshotbuilder.query.TableNameGenerator;
+import bio.terra.service.snapshotbuilder.query.TablePointer;
+import bio.terra.service.snapshotbuilder.query.TableVariable;
+import bio.terra.service.snapshotbuilder.query.filtervariable.BinaryFilterVariable;
+import bio.terra.service.snapshotbuilder.query.filtervariable.BooleanAndOrFilterVariable;
+import bio.terra.service.snapshotbuilder.query.filtervariable.FunctionFilterVariable;
+import bio.terra.service.snapshotbuilder.query.filtervariable.NotFilterVariable;
+import bio.terra.service.snapshotbuilder.query.filtervariable.SubQueryFilterVariable;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.arrow.util.VisibleForTesting;
+
+public class CriteriaQueryBuilder {
+  public static final String PERSON_ID_FIELD_NAME = "person_id";
+  private final String rootTableName;
+  private final TableNameGenerator tableNameGenerator;
+  final Map<String, TableVariable> tables = new HashMap<>();
+
+  private static final Map<String, OccurrenceTable> DOMAIN_TO_OCCURRENCE_TABLE =
+      Map.of(
+          "Condition", new OccurrenceTable("condition_occurrence", "condition_concept_id"),
+          "Measurement", new OccurrenceTable("measurement", "measurement_concept_id"),
+          "Visit", new OccurrenceTable("visit_occurrence", "visit_concept_id"),
+          "Procedure", new OccurrenceTable("procedure_occurrence", "procedure_concept_id"),
+          "Observation", new OccurrenceTable("observation", "observation_concept_id"),
+          "Device", new OccurrenceTable("device_exposure", "device_concept_id"),
+          "Drug", new OccurrenceTable("drug_exposure", "drug_concept_id"));
+
+  private static OccurrenceTable getOccurrenceTableFromDomain(String domain) {
+    OccurrenceTable occurrenceTable = DOMAIN_TO_OCCURRENCE_TABLE.get(domain);
+    if (occurrenceTable == null) {
+      throw new BadRequestException(String.format("Domain %s is not found in dataset", domain));
+    }
+    return occurrenceTable;
+  }
+
+  public CriteriaQueryBuilder(String rootTableName, TableNameGenerator tableNameGenerator) {
+    this.rootTableName = rootTableName;
+    this.tableNameGenerator = tableNameGenerator;
+    TablePointer tablePointer = TablePointer.fromTableName(rootTableName, tableNameGenerator);
+    TableVariable tableVariable = TableVariable.forPrimary(tablePointer);
+    tables.put(rootTableName, tableVariable);
+  }
+
+  private TablePointer getRootTablePointer() {
+    return getRootTableVariable().getTablePointer();
+  }
+
+  private TableVariable getRootTableVariable() {
+    return tables.get(rootTableName);
+  }
+
+  @VisibleForTesting
+  public BooleanAndOrFilterVariable generateFilterForRangeCriteria(
```

Comment:
In the case where a method is tested directly by its test class (and is therefore in the same package) I wouldn't make the method public. Using package access instead allows it to be visible to the test without making it visible to all other code.

And since package access is commonly used for that purpose, I also wouldn't use `@VisibleForTesting`; my thinking is that this is more for cases where a method is called in an unexpected way to make a test work, or when its access must be `public`. But this may not be the current TDR convention
```suggestion
 BooleanAndOrFilterVariable generateFilterForRangeCriteria(
```